StepCounterKotlin/app/src/main/java/com/google/android/gms/fit/samples/stepcounter/MainActivity.kt

```diff
@@ -0,0 +1,284 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.android.gms.fit.samples.stepcounter
+
+import android.Manifest
+import android.content.Intent
+import android.content.pm.PackageManager
+import android.graphics.Color
+import android.net.Uri
+import android.os.Build
+import android.os.Bundle
+import android.provider.Settings
+import android.view.Menu
+import android.view.MenuItem
+import android.view.View
+import androidx.appcompat.app.AppCompatActivity
+import androidx.core.app.ActivityCompat
+import androidx.core.widget.TextViewCompat
+import com.google.android.gms.auth.api.signin.GoogleSignIn
+import com.google.android.gms.fit.samples.common.logger.Log
+import com.google.android.gms.fit.samples.common.logger.LogView
+import com.google.android.gms.fit.samples.common.logger.LogWrapper
+import com.google.android.gms.fit.samples.common.logger.MessageOnlyLogFilter
+import com.google.android.gms.fitness.Fitness
+import com.google.android.gms.fitness.FitnessOptions
+import com.google.android.gms.fitness.data.DataType
+import com.google.android.gms.fitness.data.Field
+import com.google.android.material.snackbar.Snackbar
+
+const val TAG = "StepCounter"
+
+/**
+ * This sample demonstrates combining the Recording API and History API of the Google Fit platform
+ * to record steps, and display the daily current step count. It also demonstrates how to
+ * authenticate a user with Google Play Services.
+ */
+class MainActivity : AppCompatActivity() {
+    private val fitnessOptions = FitnessOptions.builder()
+            .addDataType(DataType.TYPE_STEP_COUNT_CUMULATIVE)
+            .addDataType(DataType.TYPE_STEP_COUNT_DELTA)
+            .build()
+
+    // Two maps are used in conjunction with the OAuth flow. This is in order to identify what
+    // function to execute once the flow has successfully completed.
+    private val signInActionMap = mapOf<() -> Any, Int>(
+            ::subscribe to 1,
+            ::readData to 2
+    )
+    private val signInResponseMap = signInActionMap.entries.associate { (k, v) -> v to k }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        // This method sets up our custom logger, which will print all log messages to the device
+        // screen, as well as to adb logcat.
+        initializeLogging()
+
+        checkPermissionsAndRun(::subscribe)
+    }
+
+    private fun checkPermissionsAndRun(signInAction: () -> Any) {
+        if (!requiresActivityRecognitionPermission() || hasRuntimePermission()) {
```

Comment:
hasRuntimePermission() -> permissionApproved()
I would also put it first in the OR statement

Reply:
Actually, a second comment which you could apply to the other samples. Why does this Q or not need to be separate, can't it be part of the permissionsApproved(). It makes this statement much harder to read. I would copy/paste the code I wrote for the location permission codelab and use that (link below). It hides the Q stuff... of course you want to rename it since this isn't foreground... it uses another method above it as a helper. Let me know if you have any questions.

https://github.com/googlecodelabs/while-in-use-location/blob/master/complete/src/main/java/com/example/android/whileinuselocation/MainActivity.kt#L219